spotbugs/src/main/java/edu/umd/cs/findbugs/detect/InefficientReplaceAll.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Contributions to SpotBugs
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+package edu.umd.cs.findbugs.detect;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.Code;
+import org.apache.bcel.classfile.Method;
+
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StatelessDetector;
+import edu.umd.cs.findbugs.SystemProperties;
+import edu.umd.cs.findbugs.ba.ClassContext;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+import edu.umd.cs.findbugs.classfile.MethodDescriptor;
+
+/**
+ * Find occurrences of replaceAll(String regex, String replacement) without any
+ * special regex characters.
+ *
+ * @author Lasse Lindqvist
+ */
+public class InefficientReplaceAll extends OpcodeStackDetector implements StatelessDetector {
+    private static final boolean DEBUG = SystemProperties.getBoolean("ira.debug");
+
+    private static final Logger LOGGER = Logger.getLogger(InefficientReplaceAll.class.getName());
+
+    private static final List<MethodDescriptor> methods = Collections.singletonList(
+            new MethodDescriptor("", "replaceAll", "(Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;"));
+
+    private final BugAccumulator bugAccumulator;
+
+    private static final Pattern REGEX_CHARS_PATTERN = Pattern.compile("[\\-\\.\\\\\\[\\]\\{\\}\\(\\)\\<\\>\\*\\+\\=\\?\\^\\$\\|]");
```

Comment:
Wow, that's a lot of backslashes.

You really don't need most of them inside a character class. The only characters that retain their significance in that context are the closing square bracket `]`, the backslash itself, and the hyphen - but the hyphen is still OK when it's at the start of the character class as this one is. Even an opening square bracket is taken as a literal.

Reply:
The reason was that the suggested regex broke because, if I remember correctly, the dot required backslashes. But I will test all of those independently so there will be no extra backslashes.

Reply:
This is now using
`Pattern.compile("[" + Pattern.quote("-.\\[]{}()<>*+=?^$|") + "]")`
which is quite clear.